File: java-code/src/BFS/binary_tree_level_order_traversal_II_107.java
package BFS;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class binary_tree_level_order_traversal_II_107 {

	public static List<List<Integer>> levelOrderBottom(TreeNode root) {
		Queue<TreeNode> q = new LinkedList<>();

		List<List<Integer>> return_list = new LinkedList<List<Integer>>();

		if (root == null) {
			return return_list;
		}

		q.add(root);
		
		while(!q.isEmpty()) {
			int size = q.size();
			List<Integer> temp_list = new LinkedList<Integer>();
			
			for(int i = 0; i < size; i++) {
				TreeNode currentNode = q.poll();
				temp_list.add(currentNode.val);
				
				if(currentNode.left != null)
					q.offer(currentNode.left);
				if(currentNode.right != null)
					q.offer(currentNode.right);
			}
			return_list.add(0, temp_list);
		}
		return return_list;
		
		
		
	}

	public static void main(String[] args) {
		TreeNode root = new TreeNode(12);
		root.left = new TreeNode(7);
		root.right = new TreeNode(1);
		root.left.left = new TreeNode(9);
		root.right.left = new TreeNode(10);
		root.right.right = new TreeNode(5);

		System.out.println(levelOrderBottom(root));
	}

}

File: python-code/binary_tree_level_order_traversal_ii_107.py
from typing import Optional, List

class TreeNode:
  def __init__(self, val):
    self.val = val
    self.left, self.right = None, None


def levelOrderBottom(root):
    queue = []

    queue.append(root)

    list = []

    while(len(queue) != 0):
        size = len(queue)

        temp = []
        for i in range(size):
            current = queue.pop(0)
            temp.append(current.val)

            if(current.left != None):
                queue.append(current.left)
            if(current.right != None):
                queue.append(current.right)

        list.insert(0, temp)

        print('\n')

    print(list)
    return -1



def main():
    root = TreeNode(12)
    root.left = TreeNode(7)
    root.right = TreeNode(1)
    root.left.left = TreeNode(9)
    root.right.left = TreeNode(10)
    root.right.right = TreeNode(5)
    levelOrderBottom(root)


main()

